feat(flowcontrol): Implement the FlowRegistry

LukeAVanDrie opened the pull request:
This PR introduces the complete, concrete implementation of the `FlowRegistry`, the stateful control plane for the flow control system. This is a foundational architectural component that manages the lifecycle of all flows, queues, and policies, providing a sharded, concurrent-safe view of its state to the `FlowController` workers.

The architecture is designed to prioritize data path performance and strict correctness for control plane state transitions, resulting in a robust, scalable, and maintainable foundation for the flow control engine.

This tracks #674

### Architectural Highlights

The design introduces a clear separation between the control plane (`FlowRegistry`) and the data plane (`registryShard`), employing several patterns to ensure correctness and performance under high concurrency:

*   **Serialized Control Plane (Actor Model):** The `FlowRegistry` uses an actor-like pattern. A single background goroutine processes all state change events from a channel. This serializes all mutations to the registry's core state (like scaling or GC), eliminating a significant class of race conditions.

*   **Sharded Data Plane with Fine-Grained Locking:** The registry's state is partitioned across multiple `registryShard` instances. Each shard uses **fine-grained, per-priority-band locks**, allowing concurrent data path operations across different priorities and dramatically reducing lock contention.

*   **Asynchronous, Lock-Free Signaling:** A lock-free atomic state machine is used for signaling between the data path and the control plane (e.g., for queue empty/non-empty transitions). This **completely decouples the data path from control plane backpressure**, guarantees strictly ordered signals, and prevents lost transitions even under high contention by coalescing signals at the source.

*   **"Trust but Verify" Garbage Collection:** A periodic, **time-based scanner** manages the lifecycle of idle flows. It uses a "Trust but Verify" pattern: it identifies candidate flows using an eventually-consistent cache ("Trust"), then performs a "stop-the-world" live check on the relevant priority band across all shards ("Verify") before deletion. This provides strong consistency precisely when needed while minimizing data path disruption.

*   **Immutable Flow Identity:** The `FlowKey` (ID + Priority) is immutable. To change the priority of traffic, a caller simply registers a new flow. The old flow is gracefully and automatically garbage collected once it becomes idle. This elegant design completely avoids complex and error-prone state migration logic.

### Suggested Review Path

1.  Start with the `contracts/` directory to understand the high-level interfaces and public API contracts.
2.  Next, please read the **comprehensive package documentation** (in `registry/doc.go`). This file contains the detailed architectural overview, including the concurrency and garbage collection strategies.
3.  Review the implementation files in a logical order: `shard.go` (the data plane slice), `managedqueue.go` (the stateful decorator with its lock-free signaler), `flowstate.go` (the GC cache), and finally `registry.go` (the orchestrator).
4.  Finally, review the configuration and validation logic in `config.go`.

### Testing Philosophy and Validation

This PR includes a complete and robust test suite that provides extremely high confidence in the correctness of this complex, concurrent system. The testing strategy is a key feature of this contribution:

*   **Deterministic Asynchronous Testing:** The primary tests for the `FlowRegistry` (`registry_test.go`) use a **test harness with an "event tap"**. This allows for gray-box testing of the actor model, enabling fast, deterministic, and race-free validation of asynchronous operations without sleeps or polling.
*   **Targeted Concurrency Tests:** Dedicated concurrency tests (`Test..._Concurrency_...`) exist to validate thread-safety under stress, specifically targeting the most critical race conditions like the GC/scaling lock interaction, the draining state transition, and the lock-free signaling mechanism.
*   **Isolated Unit Tests:** Lower-level components (`config`, `flowstate`, etc.) are tested in strict isolation to exhaustively validate their specific logic, error paths, and invariants.
*   **Comprehensive Coverage:** The suite achieves near-100% statement coverage on all key components and provides high behavioral coverage for the entire package.

LukeAVanDrie commented:
/assign @kfswain

ahg-g commented on pkg/epp/flowcontrol/registry/gc.go:
```diff
@@ -0,0 +1,114 @@
+// start begins a new GC timer for a given flow and generation. If a timer already exists for the flow, it is implicitly
+// stopped and replaced. This is the desired behavior, as a new call to start a timer for a flow (e.g., because it just
+// became idle) should always supersede any previous timer.
+func (gc *gcTracker) start(flowID string, generation uint64, timeout time.Duration) {
```
I expected this to be called to reset the timer when adding a new item to the flow's queue, I can't find that. If the caller logic is not part of this Pr, why do we have this file in this PR?

LukeAVanDrie replied:
Also via `applyFlowSynchronizationLocked` --> `evaluateFlowGCStateLocked`

```
// evaluateFlowGCStateLocked is the single source of truth for deciding whether to start or stop a flow's inactivity GC
// timer.
// It expects the registry's write lock to be held.
func (fr *FlowRegistry) evaluateFlowGCStateLocked(state *flowState) {
  // GC evaluation only considers the state on active shards. Draining shards do not count towards activity.
  if state.isIdle(fr.activeShards) {
    fr.logger.V(logging.DEBUG).Info("Flow is now inactive globally, starting GC timer", "flowID", state.spec.ID,
	    "timeout", fr.config.FlowGCTimeout, "generation", state.generation)
    fr.gc.start(state.spec.ID, state.generation, fr.config.FlowGCTimeout)
  } else {
    // We stop the timer unconditionally if the flow is active. The `gcTracker` handles the case where no timer is
    // running. We do not log this as it happens frequently on the hot path.
    fr.gc.stop(state.spec.ID)
  }
}
```

Outside of the public APIs (which will not be used until registry is wired into EPP), there are no unused functions or types in this package. I've tried to arrange the registry.go file logically to colocate the GC and state verification logic.

LukeAVanDrie replied:
Some of your confusion here may be regarding how GC events are fired and handles. 

This is the concurrency model:

1. Serialized Control Plane (Actor Model): The `FlowRegistry` utilizes an Actor-like pattern for its control plane. A single background goroutine processes all state change events (e.g., GC timers, queue emptiness signals) from a channel. This serializes all mutations to the registry's core state, eliminating a significant class of race conditions and simplifying the complex logic of distributed garbage collection and flow lifecycles.

2. Coarse-Grained Admin Lock: A single `sync.Mutex` on the `FlowRegistry` protects administrative operations (e.g.,`RegisterOrUpdateFlow`). This lock is acquired both by external callers and the internal event loop, ensuring
that these complex, multi-step state changes appear atomic.

3. Shard-Level R/W Lock: Each `registryShard` uses a `sync.RWMutex` to protect its internal maps. This allows multiple `controller.FlowController` workers to read from the same shard in parallel.

4. Lock-Free Data Path (Atomics): All statistics (queue length, byte size) at all layers are implemented using `atomic.Uint64`. This allows for high-performance, lock-free updates on the "data plane" hot path, where every
request modifies these counters.

This is a high throughput system. We cannot afford for the `managedQueue` to acquire locks on the data path (`Add`, `Remove`, etc.). I solve this in two ways. Closure-based callbacks with atomic statistics for lock-free stats propagation (#4 from above). Exactly-once, edge-triggered state transition events asynchronously delivered to the `FlowRegistry`'s events channel for processing (#1 from above). This allows the queue to report its state without blocking the data path and defer all state verification and lock acquisition to the registry itself.

ahg-g replied:
There is no confusion, I missed the registry.go file that was collapsed and I didn't see during the review

ahg-g replied:
It seems we are designing the system to handle a high rate of create/update/delete of flows, and this is a major source of complexity in this PR; I am not sure this is necessary at this point, we don't have the requirements for it yet

ahg-g replied:
On update: the question is whether we need it in the first place given our discussion on the data model, if a flow is a grouping mechanism within a band, and it can be exist across bands, then there is no need for migration.

On GC, we need garbage collection for sure (I proposed that if you recall, so naturally I agree it is needed), but GC can be implemented differently. We don't need to be super accurate with our tracking. We don't need timers and an event drive system to handle that, here is another approach: for each flow we maintain a timestamp of the last time it was active, a gc goroutine would periodically check all last active timestamps and GC the ones that were not active for a period of time, this will require using locks, but we should still be able to do it efficiently.

I am mainly trying to explore ways to simplify the design, but I will leave the decision to you on how to move forward since this is already implemented ...

LukeAVanDrie replied:
I'm circling back to this foundational discussion on the garbage collection architecture. The in-depth debate we had here was incredibly valuable and has directly led to a significant evolution of the design. I want to thank @ahg-g especially for pushing on the simplicity, scalability, and resource overhead aspects—it forced a re-evaluation that resulted in a much stronger, more robust final model.

The new design, which is now implemented in the PR, is a **Generational Mark-and-Sweep** model. It represents a synthesis of the best ideas from both the original event-driven model and the proposed polling-based alternative. It successfully combines the simplicity and low resource overhead of a polling loop with the performance and scalability of an event-driven system.

To summarize how this new architecture addresses all the points we've discussed, here is a comprehensive comparison of the trade-offs:

| Aspect | Old Event-Driven (Timers) | Proposed "Naive" Polling | **New Mark-and-Sweep (Final Design)** |
| :--- | :--- | :--- | :--- |
| **Hot Path Impact** | **Good.** Low overhead (non-blocking signal send) on `Add`/`Remove`. | **Poor.** Requires a **write lock on the central registry** to update `lastActiveTimestamp` on every `Add`, creating a global bottleneck. | **Excellent.** `Add`/`Remove` only interacts with lock-free atomics. **Zero hot-path lock contention.** |
| **GC Scalability** | **Good.** `O(changes)`. Work is proportional to state transitions. | **Poor.** `O(N)`. The GC loop constantly scans all `N` flows, creating CPU churn and lock contention that worsens with scale. | **Excellent.** `O(changes)`. The periodic loop does **zero work** unless the event-driven cache indicates a flow *might* be idle. It combines the best of both. |
| **Resource Overhead** | **Poor.** Requires a `time.AfterFunc` **timer per idle flow**. This is a significant resource cost and scheduling burden at scale. | **Excellent.** Only a single, periodic goroutine. No per-flow overhead. | **Excellent.** Only a single, periodic goroutine. No per-flow overhead. |
| **Complexity Location** | **Contained.** All timer and state logic lives in the "cold path" control plane (`gc.go` and `registry.go`). | **Leaky.** Moves complexity (write locks, timestamp management) directly into the **hot path**, making performance analysis harder. | **Perfectly Contained.** All complex logic (generational checks, batched locking) lives in the **cold path** (`onG...

ahg-g commented on pkg/epp/flowcontrol/registry/shard.go:
```diff
@@ -270,17 +274,190 @@ func (s *registryShard) Stats() contracts.ShardStats {
 }
 
-var _ contracts.RegistryShard = &registryShard{}
+//  --- Internal Administrative/Lifecycle Methods (called by `FlowRegistry`) ---
```
who is calling those four new private functions in this PR?

LukeAVanDrie commented on pkg/epp/flowcontrol/registry/flowstate.go:
```diff
@@ -0,0 +1,150 @@
+
+// newFlowState creates the initial state for a newly registered flow.
+// It initializes the state based on the current set of active and draining shards.
+func newFlowState(spec types.FlowSpecification, allShards []*registryShard) *flowState {
```
`applyFlowSynchronizationLocked`. Whenever there is an update that involves creating a new flow instance, we update the flow state cache.

```
// Get the combined list of all shards for state updates and propagation.
allShards := fr.getAllShardsLocked()

if !exists {
  // This is a new flow.
  state = newFlowState(spec, allShards)
  fr.flowStates[flowID] = state
} else {
  // This is an update to an existing flow.
  if state.spec.Priority != spec.Priority {
    isPriorityChange = true
    oldPriority = state.spec.Priority
  }
  state.update(spec, allShards)
}
```

LukeAVanDrie commented on pkg/epp/flowcontrol/registry/managedqueue.go:
```diff
@@ -138,59 +150,37 @@ func (mq *managedQueue) Remove(handle types.QueueItemHandle) (types.QueueItemAcc
 	cleanedItems, err = mq.queue.Cleanup(predicate)
-	if err != nil || len(cleanedItems) == 0 {
-		return cleanedItems, err
+	if err != nil {
```
From the `framework.SafeQueue` definition which managedqueue decorates.

```
// Cleanup iterates through the queue and atomically removes all items for which the predicate returns true,
// returning them in a slice.
//
// The handle for each removed item MUST be invalidated.
Cleanup(predicate PredicateFunc) (cleanedItems []types.QueueItemAccessor, err error)

// Drain atomically removes all items from the queue and returns them in a slice.
//
// The handle for all removed items MUST be invalidated. The queue MUST be empty after this operation.
Drain() (drainedItems []types.QueueItemAccessor, err error)
```

`Cleanup` takes a predicate function and does a full queue scan. We use this for removing expired items (whether from context cancellation or TLL expiry). We could trigger `Drain` behavior by passing a predicate that always returns `true` to `Cleanup`.  I thought is was semantically cleaner to separate these though. You can also write more efficient `Drain` implementations this way.

---

As an aside, eventually, I would like to support dynamic intra-flow policy updates for flows. If this involves changing the underlying comparator for a priority queue, we must atomically migrate the queue via `Drain` and multiple `Add`. I'm strongly considering removing the error return value from the `Add` and `Drain` contracts so that this update strategy can remain infallible (allowing our flow updates to stay synchronous and consistent across shards). Having `Drain` separate from `Cleanup` means this change doesn't affect the `Cleanup` error contract which still benefits from being able to return an error at the moment.

LukeAVanDrie replied:
Summarizing the outcome of this discussion: while the original motivation for `Drain` was forward-looking, it provides a clear, immediate performance benefit for graceful controller shutdown, especially for heap-based queues where it avoids an expensive re-heapify operation.

I am keeping the distinct `Cleanup` (for conditional housekeeping) and `Drain` (for unconditional evacuation) methods, as the contract and implementations are already merged and provide this tangible benefit.

LukeAVanDrie commented on pkg/epp/flowcontrol/registry/config.go:
```diff
@@ -47,50 +64,93 @@ type Config struct {
+// validateAndApplyDefaults checks the configuration for validity and mutates it to populate any empty fields with
+// system defaults. It ensures that all priority bands are well-formed, have unique priority levels and names, and that
+// their chosen plugins are compatible. This method should be called once by the registry before it initializes any
+// shards.
```
Top of the registry constructor.

```
// NewFlowRegistry creates and initializes a new `FlowRegistry` instance.
func NewFlowRegistry(
  config *Config,
  initialShardCount uint,
  logger logr.Logger,
  opts ...RegistryOption,
) (*FlowRegistry, error) {
  if err := config.validateAndApplyDefaults(); err != nil {
    return nil, fmt.Errorf("master configuration is invalid: %w", err)
  }
  ...	
```

LukeAVanDrie commented on pkg/epp/flowcontrol/registry/gc.go:
```diff
@@ -0,0 +1,114 @@
+	}
+
+	// We use AfterFunc which works efficiently with both `RealClock` and `FakeClock` (for tests).
+	timer := gc.clock.AfterFunc(timeout, func() {
```
@ahg-g & @kfswain 

We have a potential (ephemeral) goroutine leak here.

**Cause:** The `gcTracker` uses `time.AfterFunc`, which creates a new `time.Timer` and an associated goroutine for every timer started. While `Stop()` is called, the Go runtime doesn't guarantee the immediate cleanup of the underlying resources if the timer has already fired.

**Impact:** This is a very minor performance concern that only matters at O(hundreds of thousands) of flows. If the system creates and destroys tons of short-lived flows, this could lead to a build-up of dormant timer-related goroutines. 

**Solution:** For extreme hardening, we can consider replacing the one-goroutine-per-timer model with a single "timer wheel" (circular buffer) or a min-heap-based scheduler that manages all pending expirations in one goroutine. 

For most practical purposes, the current implementation is simpler and likely sufficient though.

I was wondering if this is worth documenting somewhere (Github issue, code comment, etc.) or if we just let this slide.

shmuelk replied:
This is actually a request rate times the amount of time the EPP is active problem, which might be significant.

LukeAVanDrie commented on pkg/epp/flowcontrol/registry/registry.go:
```diff
@@ -0,0 +1,848 @@
+	totalByteSize atomic.Uint64
+	totalLen      atomic.Uint64
+	// perPriorityBandStats stores *bandStats, keyed by priority (uint).
+	perPriorityBandStats sync.Map
```
This map is populated once at initialization and then only read from in the `Stats()` method. 

We could consider replacing with a standard map protected by the registry's mu lock (or an RWMutex for read access) which is more conventional and potentially more performant as `sync.Map` is optimized for a different access pattern (concurrent reads and writes).

This would simplify the code a bit as well.

The only writes we are concerned with are key additions and deletions. The values are already `atomic.Uint64`.

LukeAVanDrie replied:
Closing the loop on this thread: The design has since evolved to use a simple `map[uint]*bandStats`.
The map's keyset is immutable after initialization, and the values (`*bandStats`) contain atomic counters, which are updated lock-free. This approach is simpler and more idiomatic than the original `sync.Map` for our specific access pattern.

LukeAVanDrie commented on pkg/epp/flowcontrol/registry/registry.go:
```diff
@@ -0,0 +1,848 @@
+//
+// (See package documentation in `doc.go` for the full overview of the Architecture and Concurrency Model)
+type FlowRegistry struct {
+	config *Config
```
The `Config` is partitioned and distributed to shards. The `FlowRegistry` holds the master config, and shards hold pointers to their partitioned version.

If we ever want to support configuration hot-reloading we could consider treating the `Config` object as deeply immutable after its initial validation. The `FlowRegistry` could hold its master config within a `sync/atomic.Value`. This would provide a clear, thread-safe mechanism for updating the entire configuration atomically.; however, the current locking model is sufficient for the present design.

Is this a story we would ever want to support? There are some cascading impacts from changing the config. E.g., if you change the default inter/intra flow policies, how should we treat existing queues? This feels like an indirect way of calling `RegisterOrUpdateFlow`. I can see some cases where modifying the capacity limits, event buffer size, etc. could be useful and non-disruptive on the fly though.

LukeAVanDrie replied:
This was an exploration of a potential future feature (hot-reloading).

The final design decision for the MVP is to treat the `Config` as immutable after startup. A proper hot-reloading feature would require a dedicated `UpdateConfig` method with a well-defined strategy for applying changes to live components.

Let's defer this advanced functionality. We can create a dedicated tracking issue for it if it becomes a requirement in the future.

LukeAVanDrie commented on pkg/epp/flowcontrol/registry/registry.go:
```diff
@@ -0,0 +1,848 @@
+	if currentTotalShardCount != initialTotalShardCount {
+		// Rare race condition: Shard scaling occurred concurrently.
+		// Instead of re-preparing everything, we optimize by adjusting the existing `queues` slice.
+		fr.logger.V(logging.DEBUG).Info("Total shard count changed during preparation, adjusting queues under lock",
```
The logic to handle a shard count change during the "prepare" phase of `RegisterOrUpdateFlow` is somewhat complex. It attempts to patch the slice of prepared queues by either appending new ones or truncating the slice.

A simpler, albeit potentially less performant, approach is to wrap the prepare-commit logic in a loop. If the shard count changes, the entire operation is retried. This may be worthwhile for maintainability.

Shard scaling is likely an infrequent event compared to flow registration. The cost of re-running the preparation phase in the rare event of a race might be an acceptable trade-off for removing the complex slice manipulation logic and simplifying a critical and complex function. The current implementation is a valid optimization, so this is a question of complexity vs. performance trade-offs.

An even simpler (and even less performant) option is to move the entire prepare-commit logic under the registry lock, removing the race condition entirely.

Open to any of these options.

LukeAVanDrie commented on pkg/epp/flowcontrol/registry/shard.go:
```diff
@@ -27,44 +27,65 @@ import (
+// and the actual queue instances (`managedQueue`) for its assigned partition. It provides a read-optimized, operational
+// view for a single `controller.FlowController` worker.
+//
+// The registryShard is deliberately kept simple regarding coordination logic; it relies on the parent `FlowRegistry`
```
This is to clearly align `registryShard` internal struct with the public interface definition `contracts.RegistryShard` in `contracts/registry.go` which the `controller.shardProcessor` takes a dependency on. Of course, within the registry package, simply `shard` is probably more idiomatic.

This is a simple find and replace if you prefer this convention.

LukeAVanDrie replied:
I'll stick with `registryShard` for now for the cross-package clarity but will keep this in mind for future components to ensure we're aligned with the project's idiomatic style.

ahg-g commented on pkg/epp/flowcontrol/registry/registry.go:
```diff
@@ -0,0 +1,848 @@
+	}
+
+	for i, shard := range allShards {
+		shard.synchronizeFlow(spec, policy, queues[i])
```
name is not intuitive, `addOrUpdateFlow`?

LukeAVanDrie replied:
That's a great point. I agree that `synchronizeFlow` isn't as intuitive as it could be, and I appreciate the suggestion of `addOrUpdateFlow`.

My hesitation with `addOrUpdateFlow` is that it feels a bit like a simple CRUD operation, and doesn't fully capture the more complex state transitions the function handles, like reactivating a draining queue. My original idea was `reconcileFlow`, but I rejected it because "Reconcile" has a very specific, loaded meaning in the context of Kubernetes controllers.

What do you think of a third option: `applyFlowSpec`?

The verb "Apply" feels like a good fit here. It's well-understood to mean "take this specification and make the live state match," which is exactly what this infallible function does. It seems to strike a better balance between being intuitive and being architecturally precise than either `synchronizeFlow` or `addOrUpdateFlow`.

I'm happy to go with `addOrUpdateFlow` if you still feel it's the clearest.

ahg-g replied:
IMO the function name should focus on capturing the what not the how; but I don't want us to spend time changing this now.

ahg-g commented on pkg/epp/flowcontrol/registry/registry.go:
```diff
@@ -0,0 +1,848 @@
+		shard.synchronizeFlow(spec, policy, queues[i])
+	}
+
+	// If this was a priority change, attempt to GC the newly-draining queue immediately if it's already empty.
```
Do we need this if the default GC mechanism will clean this up anyway? this looks out of place here.

ahg-g replied:
This highly depends on the flows rate of change, it seems we are assuming it to be high, why is that?

LukeAVanDrie replied:
This entire optimization is no longer relevant with the new composite `FlowKey` design, which treats each `(ID, Priority)` pair as a distinct, independent flow. There is no longer a special "priority change" or "migration" event; the old flow simply becomes Idle as traffic shifts, and the new generational GC handles it cleanly and efficiently.
This simplification completely resolves the concern about designing for a high rate of change. The new model is much simpler.

ahg-g commented on pkg/epp/flowcontrol/registry/flowstate.go:
```diff
@@ -0,0 +1,150 @@
+	s.generation++ // Invalidate any pending GC timers for the old generation.
+
+	// If priority did not change, there's nothing more to do.
+	if oldPriority == spec.Priority {
```
I would implement this as a delete then add flow rather than having the internal flow management support update. 

However, the question in my mind is this: now that flow id is set by the requests and is decoupled from criticality definition, what happens if requests constantly use the same flow id across different criticalities?

kfswain replied:
Discussed offline, our current direction will work. Will reevaluate as we implement Phase 2 of API changes

kfswain commented on pkg/epp/flowcontrol/contracts/registry.go:
```diff
@@ -14,18 +14,147 @@ See the License for the specific language governing permissions and
+	//
+	// # Dynamic Update Behaviors
+	//
+	//   - Priority Changes: If a flow's priority level changes, its current active `ManagedQueue` instance is marked
```
SG

LukeAVanDrie replied:
In short: Is criticality a set of service tiers within a tenant or is it a set of service tiers for classifying tenants?

kfswain replied:
Discussed offline, our current direction will work. Will reevaluate as we implement Phase 2 of API changes

ahg-g commented on pkg/epp/flowcontrol/registry/config.go:
```diff
@@ -47,50 +64,93 @@ type Config struct {
+	// without blocking the data path, but consumes more memory.
+	//
+	// Optional: Defaults to 4096.
+	EventChannelBufferSize uint32
```
I don't expect that we will initially have cases for high flow churn.

ahg-g commented on pkg/epp/flowcontrol/registry/registry.go:
```diff
@@ -0,0 +1,848 @@
+			// This is safe because scale-down always removes shards from the end of the list, and the `allShards` slice is
+			// ordered with active shards first.
+			queues = queues[:currentTotalShardCount]
+		}
```
ok, it is fine, I missed the lock in the middle

> What are your thoughts on this trade-off? 

I would start simplest, evaluate and then optimize; but since you did this already, I don't recommend changing it.

LukeAVanDrie commented on pkg/epp/flowcontrol/registry/registry.go:
```diff
@@ -0,0 +1,848 @@
+	policy framework.IntraFlowDispatchPolicy,
+	queues []framework.SafeQueue,
+) {
+	flowID := spec.ID
```
> Granted, this breaks fairness only across priority bands. Which honestly, is an open question with how we want to handle fairness cross-band anyway.

This was not a well-solved problem anyways, even with the previous model I was operating under.

I think this is the most pragmatic short-term solution, and it offers some simplifications for the registry. I will instrument this change.

LukeAVanDrie replied:
Addressed in #1340 and some changes I will amend shortly on this PR.

shmuelk commented:
>> pkg/epp/flowcontrol/registry/registry.go: The FlowRegistry itself. The central orchestrator.

If the code is an orchestrator, why is it called a registry and not simply FlowOrchestrator?

LukeAVanDrie commented on pkg/epp/flowcontrol/registry/managedqueue.go:
```diff
@@ -205,20 +195,110 @@ func (mq *managedQueue) PeekHead() (types.QueueItemAccessor, error) { return mq.
+	// CRITICAL: Check if the queue is *already* empty at the moment it's marked as draining (or if it was already
+	// draining and empty). If so, we must immediately attempt the transition to Drained to ensure timely GC.
+	// This handles the race where the queue becomes empty just before or during being marked draining.
+	if mq.Len() == 0 {
```
That's a very valid concern about the potential for missed events in event-driven designs. The specific race condition you're highlighting here will actually be resolved by the simplifications coming with the composite flow key changes.

However, your broader question about event-driven vs. polling for GC is a crucial one. I've detailed my reasoning for sticking with the event-driven model in my [other comment](https://github.com/kubernetes-sigs/gateway-api-inference-extension/pull/1319#discussion_r2268050052). I believe the trade-offs discussed there, particularly regarding hot path performance and complexity containment, address why the current model was chosen over a polling/timestamp approach. I'd be happy to discuss it further if my reasoning there isn't clear.

LukeAVanDrie replied:
This was a critical discussion that directly influenced the final, much-improved GC architecture. Your concern about the fragility of a purely event-driven design and the risk of missing events in corner cases was spot on.

The new **Generational Mark-and-Sweep** model was designed specifically to solve this. It's a hybrid approach that gives us the best of both worlds:

1. It uses event signals (`BecameEmpty/NonEmpty`) to maintain a lightweight, eventually consistent cache of flow idleness. This is the performant, event-driven part.
2. However, it **does not trust this cache for correctness**. The actual GC decision is made by a simple, periodic polling loop (`onGCTick`) that performs a "Trust but Verify" check against the ground truth before deleting anything.

This hybrid model is fundamentally more robust. Even if an event were somehow missed, the periodic sweep would eventually correct the state. It has the performance benefits of an event-driven system but the safety and predictability of a polling system. This definitively resolves the concern about missed events.

LukeAVanDrie commented on pkg/epp/flowcontrol/contracts/registry.go:
```diff
@@ -14,18 +14,147 @@ See the License for the specific language governing permissions and
+//
+//   - Registered: A logical flow is Registered when it is known to the `FlowRegistry`. It has exactly one Active
+//     instance across all priority bands and zero or more Draining instances.
+//   - Active: A specific instance of a flow within a priority band is Active if it is the designated target for all
```
This is a single *flow instance* (i.e., a single queue instance for a logical flow). This was needed when we had the active vs. draining queue state for priority updates. With #1340, this is no longer relevant and this documentation will be updated.

LukeAVanDrie commented:
> > > pkg/epp/flowcontrol/registry/registry.go: The FlowRegistry itself. The central orchestrator.
> 
> If the code is an orchestrator, why is it called a registry and not simply FlowOrchestrator?

That is a very precise question. You've hit on a key point about the component's dual role. The name `FlowRegistry` was chosen deliberately because its **primary external-facing responsibility is to act as a stateful catalog of flow instances.** While it uses orchestration as an *internal implementation strategy*, its core purpose in the system's architecture is that of a registry.

**Registry Pattern:**

The `FlowRegistry` fits this pattern from a client's perspective:
*   **Registration:** The primary entry point is `RegisterOrUpdateFlow()`. A client explicitly registers a flow's specification.
*   **Lookup/Discovery:** Clients discover the state of the system via `Shards()`, which provides access to the registered entities.
*   **Deregistration:** It manages the removal of stale registrations via garbage collection.

Its main job is to be the single source of truth for "what flows exist and what is their configuration?"

**Orchestrator Pattern:**

The `FlowRegistry` uses this pattern *internally* to maintain its own state correctly in a concurrent environment:
*   It **orchestrates shard scaling** by telling individual `registryShard` instances to re-partition their configuration or enter a draining state.
*   It **orchestrates garbage collection** by reacting to events from queues and timers and then commanding shards to delete specific queue instances.

**Why `Registry` is the More Fitting Name:**

The orchestration is the *how*, not the *what*. It's the complex internal machinery that makes the registry robust.

We chose the name `FlowRegistry` because:

1.  **It describes the public contract.** From the outside, you interact with it as a registry.
2.  **It avoids confusion with the `FlowController`.** In our system, the `FlowController` is the component that actually orchestrates the *dispatch of user requests*. Calling this component `FlowOrchestrator` would create a significant naming collision and make it unclear which component is responsible for what part of the workflow.

This is an excellent point of feedback, though. It's clear my documentation could be more precise. I will update the GoDoc comment to clarify this distinction: that it **is a Registry** which **uses an internal Actor-based orchestrator** to manage its state.

Thank you for the sharp observation!

ahg-g commented on pkg/epp/flowcontrol/registry/registry.go:
```diff
@@ -0,0 +1,746 @@
+}
+
+// UpdateShardCount dynamically adjusts the number of internal state shards.
+func (fr *FlowRegistry) UpdateShardCount(n int) error {
```
what will trigger shard count updates? is it dynamic based on load?

LukeAVanDrie replied:
That's a great question as we don't have a dynamic autoscaler calling this for the initial launch; I expect shards count to be configured statically.

This method was designed as a complete, symmetrical primitive for managing the shard set. The scale-up half is used by `NewFlowRegistry` for the initial shard creation. I included the scale-down half to make the primitive complete, as a "create shards" function without a corresponding "remove shards" felt like an unfinished API.  Building this clean, atomic scaling primitive *now*—when the system's logic is fresh—is a small, low-cost investment that provides a powerful hook for that future work. It makes the system ready for the next evolution without requiring a major architectural refactor later.

While the dynamic scaling aspect is forward-looking, the core mechanism is foundational to the registry's lifecycle. We can certainly consider making it a private method for now if the public exposure is a concern.

LukeAVanDrie commented on pkg/epp/flowcontrol/types/flow.go:
```diff
@@ -44,23 +44,23 @@ type FlowKey struct {
+// The comparison is performed first by `Priority` (ascending, higher priority first) and then by `ID` (ascending).
 func (k FlowKey) Compare(other FlowKey) int {
-	if k.Priority > other.Priority {
+	if k.Priority < other.Priority { // Lower number means higher priority
```
I haven't instrumented this flip yet. It will need to be done across the Flow Control code in one pass. Keeping the old behavior for now. This was an inconsistency bug in my last PR introducing the FlowKey.

LukeAVanDrie commented on pkg/epp/flowcontrol/registry/lifecycle.go:
```diff
@@ -0,0 +1,56 @@
+// Component Lifecycle State Machine
+// =============================================================================
+
+// componentStatus represents the lifecycle state of a `registryShard`.
```
It used to be for queues and shards in the last pass. Now we don't need this for queues with the new GC logic. I will rename this to `shardStatus`.

LukeAVanDrie replied:
Done, thank again for catching this!

LukeAVanDrie commented on pkg/epp/flowcontrol/registry/doc.go:
```diff
@@ -14,27 +14,116 @@ See the License for the specific language governing permissions and
+//
+//  1. The caller registers a new flow instance, `F2` (`FlowKey{ID: "A", Priority: 20}`).
+//  2. The system treats `F1` and `F2` as independent entities (Immutable `FlowKey` design).
+//  3. As `F1` becomes Idle, it is automatically garbage collected. This achieves the outcome gracefully without complex
```
That's a great point, thank you for the clarification. The intent here is to illustrate the mechanism by which a caller can achieve a priority migration, not to imply an automatic, stateful migration within the registry itself. You are correct that `F1` and `F2` are treated as completely independent flows, and both could continue to receive traffic.
I'll rephrase this section to be more precise and make it clear that this pattern is how a caller can gracefully shift traffic from one priority to another, leading to the eventual GC of the old, unused flow.

LukeAVanDrie commented on pkg/epp/flowcontrol/registry/events.go:
```diff
@@ -0,0 +1,127 @@
+
+// event is a marker interface for internal state machine events processed by the `FlowRegistry`'s event loop.
+type event interface {
+	isEvent()
```
I agree it looks a bit unusual. This is a Go idiom called a "marker interface" used to create a closed set of types (or sometimes used for "sum type emulation"). The unexported method (`isEvent()`) provides a compile-time guarantee that only types defined within the `registry` package can satisfy the `event` interface. This prevents other packages from sending unexpected types on our events channel.

The primary alternative would be to use a struct that contains all possible event types, often with an `EventType` enum field:

```go
type Event struct { 
  Type EventType
  QueueEvent *queueStateChangedEvent // non-nil if Type is Queue
  ShardEvent *shardStateChangedEvent // non-nil if Type is Shard
}
```
I chose the marker interface approach because it's slightly cleaner at the consumption site (the `switch e := evt.(type)` in the `Run` loop) and avoids the overhead of checking for nil pointers in the unified `Event` struct. Both patterns are valid, but the marker interface felt more direct for our closed set of internal events.

LukeAVanDrie commented on pkg/epp/flowcontrol/registry/shard.go:
```diff
@@ -270,17 +274,190 @@ func (s *registryShard) Stats() contracts.ShardStats {
+
+// garbageCollect removes a queue instance from the shard.
+// This must be called under the shard's write lock.
+func (s *registryShard) garbageCollect(flowID string, priority uint) {
```
Update: @kfswain This was a great point about cohesion. In the latest version of the code, I have completely removed the separate `gc.go` utility. All garbage collection logic is now co-located directly within `registry.go` as a dedicated, well-documented "Garbage Collection Subsystem." This keeps all the state management logic for the `FlowRegistry` type within its own file, which I think addresses the concern you raised perfectly.

LukeAVanDrie commented on pkg/epp/flowcontrol/registry/registry.go:
```diff
@@ -0,0 +1,848 @@
+	}
+
+	// Perform compatibility check. (This check is also done during config validation, but repeated here defensively).
+	if err := validateBandCompatibility(*bandConfig); err != nil {
```
This redundant defensive check has been removed. The final design relies on the core principle that all configuration is validated once at startup (`NewConfig`). The control plane's serialization guarantees that no invalid configuration can be introduced at runtime, so this check was unnecessary.

ahg-g commented:
/approve
/lgtm
/hold

I approved in case you would like to address the comments in the followup PR.

LukeAVanDrie commented:
@ahg-g (cc: @kfswain) -- Thanks again for the LGTM and approval on the initial version.

After it was approved, I did another deep pass on the implementation with a focus on hardening the concurrency model and improving the long-term maintainability before merging. I've just pushed up the refined version for your final review.

The core logic is the same, but I've made a few significant enhancements that I believe are critical for this foundational layer:

1. Hardened Concurrency & Performance:
- I've moved the registryShard from a single coarse-grained lock to a fine-grained locking model (per-priority-band). This significantly reduces lock contention and improves data path parallelism.
- The signaling mechanism in managedQueue has been upgraded to a lock-free atomic state machine. This completely decouples the data path from control plane backpressure, guarantees no signals are lost, and makes the system far more resilient under high contention. It also significantly reduces noise on the events channel by coalescing signals when thrashing.


2. More Robust GC & Scaling:

- I've introduced an RWMutex (gcScaleLock) to explicitly synchronize the Garbage Collection and shard scaling operations, eliminating a potential TOCTOU race condition and making the system safer.
- The GC logic itself was simplified from a generational model (still "mark and sweep" -- the algorithm is the same; only the marker has changed) to a more intuitive time-based approach, which is easier to reason about.

3. Improved Testability: The test suite was overhauled to use a new harness with an "event tap," allowing for faster, fully deterministic validation of the asynchronous logic without any sleeps or flaky polling.

4. Improved Documentation Maintainability:

- While I still have detailed docs, they now follow the principle of first disclosure and information exists closer to the relevant types.

I'm much more confident in the robustness and performance of this version. Let me know what you think.

LukeAVanDrie commented on pkg/epp/flowcontrol/registry/config.go:
```diff
@@ -30,67 +31,213 @@ import (
+	// PriorityName is a human-readable name for this priority band (e.g., "Critical", "Standard").
+	// It must be unique across all priority bands in the configuration.
+	// Required.
+	PriorityName string
```
SGTM; I think this can be deferred and revisited when the component matures.

LukeAVanDrie commented on pkg/epp/flowcontrol/registry/flowstate.go:
```diff
@@ -0,0 +1,105 @@
+// The `shards` parameter defines the scope of the check (the current ground truth). We only verify the state against
+// this list because the internal `emptyOnShards` map might contain stale entries for shards that have been garbage
+// collected by the registry but not yet purged from this specific `flowState`.
+func (s *flowState) isIdle(shards []*registryShard) bool {
```
Done!

ahg-g commented:
> > How can I tell the difference between the latest commit and what I reviewed before? for the future, I prefer to not force push so that it is easier for the reviewer to diff
> 
> Split from my reflog, so it should be visible as an independent commit now. What is the repo best practices for merge? Do we squash?

We have github configured to automatically squash, so just send reviewer responses as separate commits.
